Add types for systemjs@6

joeldenning opened the pull request:
Hi there 👋. I am a current, active maintainer of SystemJS. We rewrote the whole project late last year and the typescript definitions are now out of date. I am hoping to maintain the 0.20 typescript definitions for those using systemjs@<1, but add definitions for systemjs@6 as the `latest` types on npm.

I attempted to do so by creating a `v0` directory and moving the old definitions there. I don't know if that is the correct thing to do, but saw other DefinitelyTyped packages doing that. If I missed anything there, let me know.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/systemjs/systemjs/blob/master/docs/api.md
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

weswigham commented on types/systemjs/index.d.ts:
```diff
@@ -1,388 +1,103 @@
+  // tslint:disable-next-line no-unnecessary-generics
+  function ImportFn<T extends Module>(moduleId: string, parentUrl?: string): Promise<T>;
+
+  function DeclareFn(_export: typeof ExportFn, _context: Context): Declare;
```
What are all these uppercase-named functions for? Near as I can tell, there's no `System.DeclareFn` in the public API?

joeldenning replied:
They are for reuse within this file and are not meant to be exported. I noticed that they were being exported but didn't know how to implement types, interfaces, and functions in a global typescript declaration without all of them being exported. I tried to follow [this pattern](https://www.typescriptlang.org/docs/handbook/namespaces.html#working-with-other-javascript-libraries) from the typescript handbook.

Note that my types here have a `SystemJS` namespace and then a `System` const. Newly with systemjs@6, there is actually no longer a `SystemJS` global variable at all. The only global variable is `System`.

weswigham replied:
Add `export {}` into the namespace declaration - then only other members with an `export` modifier will be exported. But if you's just using these to `typeof` them later, it's probably better if you just used type aliases, too, eg `type ImportFn = <T extends Module>(moduleId: string, parentUrl?: string) => Promise<T>;`

joeldenning replied:
👍 thanks I will update this.

joeldenning commented on types/systemjs/index.d.ts:
```diff
@@ -1,388 +1,104 @@
+   * You may optionally provide a parentUrl that will be used for resolving relative urls.
+   */
+  // tslint:disable-next-line no-unnecessary-generics
+  export const import: ImportFn;
```
Typescript fails to compile this because it thinks it's invalid to export a thing called `import`. Any ideas on how to fix?

weswigham replied:
Heh, we interestingly you actually can't have a const named import - go try it in your js repl of choice. Object members can be keywords, though. You may need to replace the namespace with a const to handle this. :S

joeldenning replied:
👍 I can switch it over to being a const instead of a namespace. How will I be able to create reusable types that don't get exported, though, if it's a const?

weswigham replied:
You can merge a namespace that only contains types with the const.

joeldenning commented on types/systemjs/index.d.ts:
```diff
@@ -1,388 +1,104 @@
+   * The returned function is intended for use after re-importing the module. Calling the function
+   * will re-bind all the exports of the re-imported module to every module that depends on the module.
+   */
+  export function delete(moduleId: string): false | UpdateModuleFn;
```
Typescript fails to compile this because it thinks it's invalid to export a thing called delete. Any ideas on how to fix?

joeldenning commented on types/systemjs/systemjs-tests.ts:
```diff
@@ -1,68 +1,81 @@
+            }
+        ],
+        execute() {
+            _export('a', 'thing');
```
The tests fail on these `_export()` calls, even though they used to be working and I have no idea why they're failing now.

The error message I'm getting:

```
Error: /Users/joeldennning/code/DefinitelyTyped/types/systemjs/systemjs-tests.ts:25:21
ERROR: 25:21  expect  TypeScript@next compile error: 
Argument of type '"a"' is not assignable to parameter of type 'never'.
ERROR: 26:21  expect  TypeScript@next compile error: 
Argument of type '"b"' is not assignable to parameter of type 'never'.
ERROR: 27:21  expect  TypeScript@next compile error: 
Argument of type '"c"' is not assignable to parameter of type 'never'.
ERROR: 28:13  expect  TypeScript@next compile error: 
Expected 2 arguments, but got 1.
```

I don't know where the `never` is coming from in that error.

joeldenning replied:
Figured this one out -- went back to polymorphic function signature and `typeof`, which solved the problem here.

weswigham replied:
Oh, it's because you don't want a union type, you want an overload list. Looks like this:
```ts
type ExportFn = {
    (exportName: string, value: any): void;
    (exports: object): void;
}; 
```

joeldenning commented:
Thanks for the bump - I have updated this with your feedback. The linter didn't like that it was a type instead of an interface, so i changed it to an interface.

joeldenning commented:
> You can merge a namespace that only contains types with the const.

How can I do this? Will it result in the namespace I create becoming a global variable itself?

weswigham commented:
```
declare var System: System.Instance;
declare namespace System {
    export interface Instance {}
    export interface Blah {}
}
```

joeldenning commented:
👍 thanks for your help! I just pushed a commit that does that.

weswigham commented on types/systemjs/index.d.ts:
```diff
@@ -1,388 +1,103 @@
+   * Loads a javascript module from either a url or bare specifier that is in an import map.
+   * You may optionally provide a parentUrl that will be used for resolving relative urls.
+   */
+  // tslint:disable-next-line no-unnecessary-generics
```
This suppression seems unused here? right?

joeldenning replied:
Yep good catch. Just pushed a commit that removes that

joeldenning commented:
🎉
